lib/logstash/outputs/kafka.rb

```diff
@@ -250,27 +250,35 @@ def retrying_send(batch)
         begin
           # send() can throw an exception even before the future is created.
           @producer.send(record)
-        rescue org.apache.kafka.common.errors.TimeoutException => e
+        rescue org.apache.kafka.common.errors.RetriableException => e
+          logger.info("KafkaProducer#send() failed, will retry", :exception => e)
           failures << record
           nil
-        rescue org.apache.kafka.common.errors.InterruptException => e
-          failures << record
-          nil
-        rescue org.apache.kafka.common.errors.SerializationException => e
-          # TODO(sissel): Retrying will fail because the data itself has a problem serializing.
-          # TODO(sissel): Let's add DLQ here.
-          failures << record
+        rescue org.apache.kafka.common.KafkaException => e
+          # This error is not retriable, drop event
+          # TODO: add DLQ support
+          logger.warn("KafkaProducer#send() failed, dropping record", :exception => e, :record_value => record.value)
           nil
         end
-      end.compact
+      end
 
       futures.each_with_index do |future, i|
-        begin
-          result = future.get()
-        rescue => e
-          # TODO(sissel): Add metric to count failures, possibly by exception type.
-          logger.warn("producer send failed", :exception => e.class, :message => e.message)
-          failures << batch[i]
+        # We cannot skip nils using `futures.compact` because then our index `i` will not align with `batch`
+        unless future.nil?
+          begin
+            future.get
+          rescue java.util.concurrent.ExecutionException => e
+            # TODO(sissel): Add metric to count failures, possibly by exception type.
+            if e.get_cause.is_a? org.apache.kafka.common.errors.RetriableException
+              logger.info("KafkaProducer#send() future failed, will retry", :exception => e.get_cause)
```

Comment:
```suggestion
              logger.info("producer send failed, will retry sending", :exception => e.cause.class, :message => e.cause.message)
```

pretty much the whole `future.get` exception handling section should get aligned with whatever we do above

Reply:
`:message => e.cause.message` as suggested or `:message => e.message` like it was?

Reply:
PTAL